src/opendr/perception/speech_recognition/quadraticselfonn/quadraticselfonn_learner.py

```diff
@@ -0,0 +1,267 @@
+            for batch_id, (x, y) in enumerate(dataloader):
+                self.optimizer_func.zero_grad()
+                output = self._get_model_output(x)
+                y = y.to(self.device)
+                loss = self.loss(output, y)
+                loss.backward()
+                self.optimizer_func.step()
+                statistics[epoch]["batch_losses"].append(loss.data.item())
+                if verbose and not silent:
+                    logging.info(f"Batch {batch_id}: training loss {loss.data.item():.7}")
+            if val_dataset is not None:
+                statistics[epoch]["validation_results"] = self.eval(val_dataset)
+                if not silent:
+                    logging.info(f"Epoch {epoch} validation results:\n"
+                                 f"Accuracy: {statistics[epoch]['validation_results']['test_accuracy']:.4}\n"
+                                 f"Total loss: {statistics[epoch]['validation_results']['test_total_loss']:.7}")
+            if not self.checkpoint_after_iter == 0 and epoch % self.checkpoint_after_iter == 0:
+                filename = os.path.join(self.temp_path + f"QuadraticSelfONN-{epoch}.pth")
+                self.save(filename)
+                if not silent:
+                    logging.info(f"Saved checkpoint to {filename}")
+
+        return statistics
+
+    def eval(self, dataset):
+        dataloader = DataLoader(dataset, batch_size=self.batch_size, pin_memory=self.device == "cuda")
+        self.model.eval()
+        test_loss = 0
+        correct_predictions = 0
+        for batch_id, (x, y) in enumerate(dataloader):
+            output = self._get_model_output(x)
+            y = y.to(self.device)
+            test_loss += self.loss(output, y).data.item()
+            predictions = output.max(1, keepdim=True)[1]
+            correct_predictions += predictions.eq(y.view_as(predictions)).sum().item()
+        return {"test_accuracy": correct_predictions / len(dataset),
+                "test_total_loss": test_loss}
+
+    def infer(self, batch):
+        self.model.eval()
+        if isinstance(batch, list):
+            data = np.vstack([series.numpy() for series in batch])
+        elif isinstance(batch, Timeseries):
+            data = batch.numpy()
+        else:
+            raise TypeError("infer requires Timeseries or list of Timeseries as input.")
+        output = self._get_model_output(data)
+        prediction = output.max(1, keepdim=True)
+        batch_predictions = []
+        for target, confidence in zip(prediction[1], prediction[0].exp()):
+            batch_predictions.append(Category(target.item(), confidence=confidence.item()))
+        return batch_predictions[0] if len(batch_predictions) == 1 else batch_predictions
+
+    def save(self, path):
+        if not os.path.isdir(path):
+            os.makedirs(path, exist_ok=True)
+
+        folder_basename = os.path.basename(path)
+        model_path = os.path.join(path, folder_basename + ".pt")

```

Comment:
There is no need to include the external path/folder name into the model_paths field of the json file, since this is already known (since we know where the model folder is).